spring-integration-core/src/main/java/org/springframework/integration/router/AbstractMappingMessageRouter.java

```diff
@@ -220,6 +258,9 @@ private void addChannelFromString(Collection<MessageChannel> channels, String ch
 		MessageChannel channel = resolveChannelForName(channelName, message);
 		if (channel != null) {
 			channels.add(channel);
+			if (!mapped && !(this.dynamicChannels.get(channelName) != null)) {
+				this.dynamicChannels.put(channelName, channel);
```

Comment:
Not sure that my comment on commit is visible for everyone, so let me share it here one more time :smile: 

M-m-m. Since we have a cache style now, I think we always should put to refresh the old entry.
Otherwise we may just route and instantly be remove by the policy on the next subsequent new route.
No? What am I missing ?